### PR TITLE
CSHARP-2672: Fix Snappy compression

### PR DESCRIPTION
**Pull request by vincentkam**
https://jira.mongodb.org/browse/CSHARP-2672
Rebased to include https://github.com/rstam/mongo-csharp-driver/pull/128

### CODE REVIEW

File: evergreen/evergreen.yml [hunk 1]
```diff
@@ -218,7 +218,7 @@ functions:
           export FLE_AWS_ACCESS_KEY_ID=${FLE_AWS_ACCESS_KEY_ID}
           export FLE_AWS_SECRET_ACCESS_KEY=${FLE_AWS_SECRET_ACCESS_KEY}
           ${PREPARE_SHELL}
-          AUTH=${AUTH} SSL=${SSL} MONGODB_URI="${MONGODB_URI}" TOPOLOGY=${TOPOLOGY} OS=${OS} evergreen/run-tests.sh
+          AUTH=${AUTH} SSL=${SSL} MONGODB_URI="${MONGODB_URI}" TOPOLOGY=${TOPOLOGY} OS=${OS} COMPRESSOR=${COMPRESSOR} evergreen/run-tests.sh
```
Comment (vincentkam) [inline evergreen/evergreen.yml:221]: This line is needed to pass the compressor variable to `run-tests.sh`

File: evergreen/evergreen.yml [hunk 2]
```diff
@@ -483,12 +483,12 @@ buildvariants:
   tasks:
      - name: "test"
 
-#- matrix_name: "tests-snappy-compression"
-#  matrix_spec: { compressor : "snappy", auth: "noauth", ssl: "nossl", version: ["3.4", "3.6", "4.0", "4.2", "latest"], topology: "standalone", os: "*" }
-#  display_name: "${version} ${compressor} ${topology} ${auth} ${ssl} ${os} "
-#  tags: ["tests-variant"]
-#  tasks:
-#     - name: "test"
+- matrix_name: "tests-snappy-compression"
```
Comment (vincentkam) [inline evergreen/evergreen.yml:486]: Re-enabling snappy compression tests

File: evergreen/run-tests.sh [hunk 3]
```diff
@@ -14,6 +14,7 @@ AUTH=${AUTH:-noauth}
 SSL=${SSL:-nossl}
 MONGODB_URI=${MONGODB_URI:-}
 TOPOLOGY=${TOPOLOGY:-server}
+COMPRESSOR=${COMPRESSOR:-none}
```
Comment (vincentkam) [inline evergreen/run-tests.sh:17]: Set the compressor to "none" if the env var does not exist

File: evergreen/run-tests.sh [hunk 4]
```diff
@@ -30,6 +31,15 @@ provision_ssl () {
   fi
 }
 
+provision_compressor () {
+    uri_environment_variable_name=$1
+    if [[ "${!uri_environment_variable_name}" =~ "/?" ]]; then
```
Comment (vincentkam) [inline evergreen/run-tests.sh:36]: I chose a different approach here than `provision_ssl`. `provision_ssl` searches for a specific combination of options that results in a URI that doesn't have additional options e.g. a URI of the form `mongodb://user:pass@localhost/`. Determing this combination of options is nontrivial, so instead, I've chosen to use a regex to search for `/?` in the URI. The downside of this the script will only run on bash instead of a generic POSIX compliant shell (e.g. whatever may be at `/bin/sh`), but this shouldn't be a problem since we were using `#/bin/bash` before anyways.

File: src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs [hunk 5]
```diff
@@ -27,28 +26,104 @@ internal class SnappyCompressor : ICompressor
...
+#endif
+        }
+
+        private int ReadInt32(Stream stream)
+        {
+            var bytesRead = new byte[4];
+            stream.ReadBytes(bytesRead, offset: 0, count: bytesRead.Length, CancellationToken.None);
+            return bytesRead[0] | (bytesRead[1] << 8) | (bytesRead[2] << 16) | (bytesRead[3] << 24);
+        }
+
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs]: not needed empty line?

File: src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs [hunk 6]
```diff
@@ -27,28 +26,104 @@ internal class SnappyCompressor : ICompressor
...
+            var uncompressedSize = ReadInt32(output);
+            output.Position = position; // return head to original position
+
+            var uncompressedMessage = new byte[uncompressedSize]; // does not include uncompressed message headers
+            input.ReadBytes(uncompressedMessage, offset: 0, count: uncompressedSize, CancellationToken.None);
+            var maxSize = SnappyCodec.GetMaxCompressedLength(uncompressedSize);
+            var compressedBytes = new byte[maxSize];
+            var compressedSize = SnappyCodec.Compress(uncompressedMessage, 0, uncompressedSize, compressedBytes, 0);
+            output.Write(compressedBytes, 0, compressedSize);
+
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs]: not necessary empty line?

**Comment by vincentkam** [comment]
Ready for review!

File: evergreen/run-tests.sh [hunk 7]
```diff
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/env bash
```
Comment (vincentkam) [inline evergreen/run-tests.sh:1]: This makes is so that the path is searched for `bash`. This is especially useful on when the user may be using newer version of bash that's not stored at `/bin/bash` (e.g. homebrew bash on MacOS)

**Comment by vincentkam** [comment]
Sorry, had to fix one more thing: now ready for review!

File: evergreen/run-tests.sh (hunk 4 above)
Comment (vincentkam) [reply evergreen/run-tests.sh:36]: This function adds the compressor to the URI, so this'll enable `zlib` testing too

**Review by DmitryLukyanov — APPROVED** [review]
Two small formatting comments and then LGTM

File: src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs (hunk 5 above)
Comment (vincentkam) [reply src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs]: Fixed!

File: src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs (hunk 6 above)
Comment (vincentkam) [reply src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs]: Fixed!

File: src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs [hunk 8]
```diff
@@ -27,28 +27,89 @@ internal class SnappyCompressor : ICompressor
...
+            output.Seek(uncompressedSizeOffset, SeekOrigin.Current); // rewind head
+            var uncompressedSize = ReadInt32(output);
+            output.Position = position; // return head to original position
+
+            var uncompressedMessage = new byte[uncompressedSize]; // does not include uncompressed message headers
+            input.ReadBytes(uncompressedMessage, offset: 0, count: uncompressedSize, CancellationToken.None);
+            var maxSize = SnappyCodec.GetMaxCompressedLength(uncompressedSize);
+            var compressedBytes = new byte[maxSize];
+            var compressedSize = SnappyCodec.Compress(uncompressedMessage, 0, uncompressedSize, compressedBytes, 0);
+            output.Write(compressedBytes, 0, compressedSize);
```
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs:45]: I don't like the idea of *reading* from `output`.

Will this work instead:

        public void Compress(Stream input, Stream output)
        {
    #if NET452 || NETSTANDARD2_0
            var uncompressedSize = (int)(input.Length - input.Position);
            var uncompressedBytes = new byte[uncompressedSize]; // does not include uncompressed message headers
            input.ReadBytes(uncompressedBytes, offset: 0, count: uncompressedSize, CancellationToken.None);
            var maxCompressedSize = SnappyCodec.GetMaxCompressedLength(uncompressedSize);
            var compressedBytes = new byte[maxCompressedSize];
            var compressedSize = SnappyCodec.Compress(uncompressedBytes, 0, uncompressedSize, compressedBytes, 0);
            output.Write(compressedBytes, 0, compressedSize);
    #else
            throw new NotSupportedException();
    #endif
        }

Also... let's try and reserve word "message" for when talking about the entire message, including headers.

File: src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs [hunk 9]
```diff
@@ -27,28 +27,89 @@ internal class SnappyCompressor : ICompressor
...
+            input.Seek(0, SeekOrigin.Begin); // rewind head
+            const int headerSize =
+                MessageLengthSize + RequestIdSize + ResponseToSize + OpCodeSize +
+                OriginalOpCodeSize + UncompressedSizeSize + CompressorTypeSize;
+            var messageLength = ReadInt32(input) - headerSize;
+            input.Position = position; // return head to original position
+            var compressedMessage = new byte[messageLength];
+            input.ReadBytes(compressedMessage, offset: 0, count: messageLength, CancellationToken.None);
+            var decompressedMessage = SnappyCodec.Uncompress(compressedMessage);
+            output.Write(decompressedMessage, offset: 0, count: decompressedMessage.Length);
```
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs]: I don't like the idea of *reading* from `output`.

Will this work instead:

            public void Decompress(Stream input, Stream output)
            {
    #if NET452 || NETSTANDARD2_0
                var compessedSize = (int)(input.Length - input.Position);
                var compressedBytes = new byte[compessedSize];
                input.ReadBytes(compressedBytes, offset: 0, count: compessedSize, CancellationToken.None);
                var decompressedBytes = SnappyCodec.Uncompress(compressedBytes);
                output.Write(decompressedBytes, offset: 0, count: decompressedBytes.Length);
    #else
                throw new NotSupportedException();
    #endif
            }

File: src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs [hunk 10]
```diff
@@ -27,28 +27,89 @@ internal class SnappyCompressor : ICompressor
...
 
+        /* Declared in header order as defined in the spec:
+         * https://github.com/mongodb/specifications/blob/7461d48df580b2e32b33fdc01188c86fc79773c1/source/compression/OP_COMPRESSED.rst#op_compressed*/
+        private const int MessageLengthSize = sizeof(int);
+        private const int RequestIdSize = sizeof(int);
+        private const int ResponseToSize = sizeof(int);
+        private const int OpCodeSize = sizeof(int);
+        private const int OriginalOpCodeSize = OpCodeSize;
+        private const int UncompressedSizeSize = sizeof(int);
+        private const int CompressorTypeSize = sizeof(byte); // compressorId in spec
```
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs]: These constants might no longer be needed depending on what changes you make below.

File: src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs [hunk 11]
```diff
@@ -27,28 +27,89 @@ internal class SnappyCompressor : ICompressor
...
+        /// <param name="input"> The input stream. Stream MUST HAVE read the exactly up to and including the
+        /// originalOpCode.</param>
+        /// <param name="output">The output stream. Stream MUST HAVE written the header exactly up to and including
+        /// compressorType aka compressorId.</param>
+        /// <remarks>
+        /// The precondition for <paramref name="input"/> mirrors the precondition for <see cref="Decompress"/>'s output
+        /// precondition.
+        /// The precondition for <paramref name="output"/> mirrors the precondition for <see cref="Decompress"/>'s input
+        /// precondition.
+        /// </remarks>
```
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs]: These comments might no longer be correct depending on what changes you make below.

File: src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs [hunk 12]
```diff
@@ -27,28 +27,89 @@ internal class SnappyCompressor : ICompressor
...
+        /// <param name="input">The input stream. Stream MUST HAVE read the header exactly up to and including
+        /// the compressorType aka CompressorId</param>
+        /// <param name="output">The output stream. Stream MUST HAVE written the message exactly up to and including
+        /// originalOpCode.</param>
+        /// <remarks>
+        /// The precondition for <paramref name="input"/> mirrors the precondition for <see cref="Compress"/>'s output
+        /// precondition.
+        /// The precondition for <paramref name="output"/> mirrors the precondition for <see cref="Compress"/>'s input
+        /// precondition.
+        /// </remarks>
```
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs]: These comments might no longer be correct depending on what changes you make below.

File: src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs [hunk 13]
```diff
@@ -27,28 +27,89 @@ internal class SnappyCompressor : ICompressor
...
+            var compressedMessage = new byte[messageLength];
+            input.ReadBytes(compressedMessage, offset: 0, count: messageLength, CancellationToken.None);
+            var decompressedMessage = SnappyCodec.Uncompress(compressedMessage);
+            output.Write(decompressedMessage, offset: 0, count: decompressedMessage.Length);
 #else
             throw new NotSupportedException();
 #endif
         }
+
+        private int ReadInt32(Stream stream)
```
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs]: Might no longer be needed.

**Review by rstam — CHANGES_REQUESTED** [review]

File: src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs (hunk 10 above)
Comment (vincentkam) [reply src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs]: Ack, fixed

File: src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs (hunk 11 above)
Comment (vincentkam) [reply src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs]: Hmm.. I think the preconditions technically still hold true, despite us changing the implementation, and leaving them there makes it easier to reason about the state of the stream (as the preconditions precisely define what the state the stream is in). I'd strongly advocate leaving them in as this was the part that was hardest to wrap my head around when I was investigating and imagine the comments will be useful in the future if we return to this section

**Review by vincentkam — COMMENTED** [review]
Ready for review again while I investigate the latest errors in EG

File: src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs (hunk 13 above)
Comment (vincentkam) [reply src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs]: Fixed!

File: tests/MongoDB.Driver.Legacy.Tests/MongoCollectionTests.cs [hunk 14]
```diff
@@ -3611,6 +3611,9 @@ private MongoCollection<BsonDocument> GetCollection(MongoServer server)
 
         private MongoServerInstance GetPrimary(MongoServer server)
         {
+            SpinWait.SpinUntil(
```
Comment (DmitryLukyanov) [inline tests/MongoDB.Driver.Legacy.Tests/MongoCollectionTests.cs]: why is this change required?

File: src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs (hunk 9 above)
Comment (vincentkam) [reply src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs]: Your suggestion is definitely cleaner and works. Fixed!

File: src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs (hunk 8 above)
Comment (vincentkam) [reply src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs:45]: Re: "message": ack. 
Your suggestion is definitely cleaner and works. Fixed!

File: src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs [hunk 15]
```diff
@@ -27,25 +27,57 @@ internal class SnappyCompressor : ICompressor
...
+        /// precondition.
+        /// </remarks>
         public void Decompress(Stream input, Stream output)
         {
-#if NET452 || NETSTANDARD2_0
-            using (var snappyStream = new SnappyStream(input, CompressionMode.Decompress, true))
-            {
-                snappyStream.CopyTo(output);
-            }
+#if NET452 || NETSTANDARD2_
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs]: Is it a typo: `2_`?

Comment (vincentkam) [reply src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs]: Thanks Dima! Fixed!

File: tests/MongoDB.Driver.Legacy.Tests/MongoCollectionTests.cs (hunk 14 above)
Comment (vincentkam) [reply tests/MongoDB.Driver.Legacy.Tests/MongoCollectionTests.cs]: I think it's because `TestFindAndRemoveWithMaxTime` spins up a new `MongoClient` by passing in `retryWrites: false`, resulting in a race condition where we try to get the primary before the topology has finished initializing. 

See: https://jira.mongodb.org/browse/CSHARP-2755 and its PR for a bit more detail
(Robert's also looking more deeply into it).

File: Release Notes/Release Notes v2.9.2.md [hunk 16]
```diff
@@ -1,7 +1,8 @@
-# .NET Driver Version 2.9.2 Placeholder Release Notes
+# .NET Driver Version 2.9.2 Release Notes
 
-The main new features in 2.9.2 are:
+This is a patch release that fixes several bugs reported since 2.9.0 was released.
```
Comment (rstam) [inline Release Notes/Release Notes v2.9.2.md]: Since 2.9.1 (not 2.9.0)?

File: src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs (hunk 11 above)
Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs]: They are no longer preconditions. They only became preconditions when you were seeking backwards into the stream to peek at the data that you assumed had been planted there.

I would like to see these comments removed.

File: src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs [hunk 17]
```diff
@@ -27,25 +27,57 @@ internal class SnappyCompressor : ICompressor
...
+        /// precondition.
+        /// </remarks>
         public void Decompress(Stream input, Stream output)
         {
 #if NET452 || NETSTANDARD2_0
-            using (var snappyStream = new SnappyStream(input, CompressionMode.Decompress, true))
-            {
-                snappyStream.CopyTo(output);
-            }
+            var compessedSize = (int) (input.Length - input.Position);
```
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs]: Type, should be `compressedSize` (missing an `r`)

**Review by rstam — CHANGES_REQUESTED** [review]

**Comment by vincentkam** [comment]
Ready for re-review! Primarily, I've changed the evergreen.yml file to drop 3.4 from testing, and updated the release notes with some snappy information. I know that typically we've done the release notes updating as a separate PR, but this seemed like an easy way to "amortize" some of the effort

File: src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs (hunk 17 above)
Comment (vincentkam) [reply src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs]: Fixed!

File: src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs (hunk 11 above)
Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs]: Think of it this way: if we had a file called `SnappyCompressorTests` (not sure why we don't) you would NOT need to write any headers to the stream before calling the `Compress` method.

So this is NOT a precondition.

Comment (vincentkam) [reply src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs]: Applied changes as discussed offline!

File: src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs (hunk 12 above)
Comment (vincentkam) [reply src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs]: Applied changes as discussed offline!

File: tests/MongoDB.Driver.Legacy.Tests/MongoCollectionTests.cs (hunk 14 above)
Comment (vincentkam) [reply tests/MongoDB.Driver.Legacy.Tests/MongoCollectionTests.cs]: This change has been removed and replaced with Robert's "Equals" fix

File: Release Notes/Release Notes v2.9.2.md (hunk 16 above)
Comment (vincentkam) [reply Release Notes/Release Notes v2.9.2.md]: Fixed!

**Comment by vincentkam** [comment]
Ready for re^2-review!

**Review by rstam — APPROVED** [review]

**Review by DmitryLukyanov — APPROVED** [review]

**Comment by vincentkam** [comment]
Rebased, squashed, date updated and pushed!
